### PR TITLE
EVM Relayer DB migrations PoC

### PR DESCRIPTION
**Pull request by krehermann**
## Ticket
 
  [BCF-3266](https://smartcontract-it.atlassian.net/browse/BCF-3266)

The point of this PR is to propose a mechanism for DB migrations that are scoped to schema.

The motivation is that (one day) LOOPPs will need to own and manage migrations.

The scope of this example is:
- Pattern for templating migrations that need to be resolved and run for multiple instances (think evm relayer per chain)
	- supports both go and sql migrations
- Isolation of the migrations so that one day the can be moved out of this repo
	- see `core/store/migrate/plugins` for all the feature work 	
- Dynamic schema creation by the resource that needs it ie the evm relayer in this case
	- there are some dirty hacks to make that happen in this PoC, but point is to illustrate the moving parts
	- in this case creating an evm relayer will run the migrations specific to that chain id 	
- CLI plumbing to manage the lifecycle of the dynamic migrations
	- in this limited example we have `node db migrate|rollback|status|version --relayer evm -chain-id <id>`

	
```
> ./chainlink node db status --relayer evm --chain-id 42
time=2024-06-18T16:05:45.964-06:00 level=INFO msg="    Applied At                  Migration"
time=2024-06-18T16:05:45.964-06:00 level=INFO msg="    ======================================="
time=2024-06-18T16:05:45.966-06:00 level=INFO msg="    Pending                  -- 0001_create_schema.tmpl.sql"
time=2024-06-18T16:05:45.967-06:00 level=INFO msg="    Pending                  -- 0002_initial.go"
> ./chainlink node db migrate --relayer evm --chain-id 42
2024-06-18T16:05:55.281-0600 [INFO]  Migrating database: "postgres://chainlink_dev:insecurepassword@localhost:5432/chainlink_development_test?sslmode=disable" cmd/shell_local.go:1008          version=2.13.0@dbd5cf7 
Generated migrations: [/var/folders/91/h_s80sjd1f3267zl7pgdlvd40000gn/T/evm_42/42/0001_create_schema.sql]
time=2024-06-18T16:05:55.321-06:00 level=INFO msg="OK   0001_create_schema.sql (5.81ms)"
time=2024-06-18T16:05:55.358-06:00 level=INFO msg="OK   0002_initial.go (37.12ms)"
time=2024-06-18T16:05:55.358-06:00 level=INFO msg="goose: successfully migrated database to version: 2"
> ./chainlink node db status --relayer evm --chain-id 42
time=2024-06-18T16:06:07.394-06:00 level=INFO msg="    Applied At                  Migration"
time=2024-06-18T16:06:07.394-06:00 level=INFO msg="    ======================================="
time=2024-06-18T16:06:07.398-06:00 level=INFO msg="    Tue Jun 18 22:05:55 2024 -- 0001_create_schema.tmpl.sql"
time=2024-06-18T16:06:07.399-06:00 level=INFO msg="    Tue Jun 18 22:05:55 2024 -- 0002_initial.go"
> ./chainlink node db version --relayer evm --chain-id 42
2024-06-18T16:06:14.150-0600 [INFO]  EVM database version: 2                            cmd/shell_local.go:1081          version=2.13.0@dbd5cf7 


```

- Testing infra for the database
	- ability to programmatically instantiate the correct db resource (ie the correct schema with the right tables)
		- this is tricky because all of the existing tests *assume* a properly configured db external to the tests
		- that can't work when the schemas are dynamic
				- if one test use evm relayer chainid=7 and other chainid=11, they have different resource requirements
		- this is solved by refactoring our txdb and introducing an evm-specific helper to instantiate the correct db schema
- Proof that something like this can work
	- we refactor the evm forwarder to use all of this machinery (this was a arbitrary choice, informed by its relatively simple ORM)
		- update the SQL statements to depend on the runtime schema
		- remove evm_chain_id from the tables
		- update test to use the evm-specific DB helper so that the correct schema is generated for the test
 	 	 	 

Out of scope
- test refactoring require pgtest to be available to the evm helper. this implies we need to put pgtest somewhere common
- anything non-evm related. but the pattern ought to work for other relayers and applications
- anything beyond the evm forwarder refactor to prove it can work.
- clean initialization code in the relayer. db migrations need a sql.DB. the relayer only has a sqlutil.Datasource. we'll need to reconcile this to support dynamic migrations (or move the dynamic migrations up the stack)

Opens:
rollback
- the dynamic migration version are decoupled from the core node version. in principle this is a gap because an operator would need to rollback each schema manually

## Resolves Dependencies

[BCF-3266]: https://smartcontract-it.atlassian.net/browse/BCF-3266?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/chains/evm/forwarders/forwarder_manager_test.go [hunk 1]
```diff
@@ -41,7 +42,8 @@ var SimpleOracleCallABI = evmtypes.MustGetABI(operator_wrapper.OperatorABI).Meth
 
 func TestFwdMgr_MaybeForwardTransaction(t *testing.T) {
 	lggr := logger.Test(t)
-	db := pgtest.NewSqlxDB(t)
+	testChainId := testutils.FixtureChainID
```
Comment (dimriou) [inline core/chains/evm/forwarders/forwarder_manager_test.go:45]: BTW you can use `github.com/smartcontractkit/chainlink/v2/core/chains/evm/testutils` to get FixtureChainID which is not coupled with Core.

File: core/store/migrate/plugins/relayer/evm/testutils/db.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/scylladb/go-reflectx"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/smartcontractkit/chainlink/v2/core/config/env"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils/pgtest"
+	"github.com/smartcontractkit/chainlink/v2/core/store/dialects"
+	"github.com/smartcontractkit/chainlink/v2/core/store/migrate/plugins/relayer/evm"
+	"github.com/smartcontractkit/chainlink/v2/internal/testdb"
```
Comment (dimriou) [inline core/store/migrate/plugins/relayer/evm/testutils/db.go:22]: Is this going to be migrated to common?

Comment (krehermann) [reply core/store/migrate/plugins/relayer/evm/testutils/db.go:22]: yes, i think we would need to mv some pgtest, testdb, dialects to common or their own module and evm code to integration repo.

Comment (jmank88) [reply core/store/migrate/plugins/relayer/evm/testutils/db.go:22]: Is this helpful? https://github.com/smartcontractkit/chainlink-common/pull/450

File: core/web/evm_forwarders_controller.go [hunk 3]
```diff
@@ -79,6 +92,13 @@ func (cc *EVMForwardersController) Delete(c *gin.Context) {
 		jsonAPIError(c, http.StatusUnprocessableEntity, err)
 		return
 	}
+	// this is a breaking change, but it's necessary to support multiple chains
+	chainId, err := stringutils.ToInt64(c.Param("chainID"))
```
Comment (reductionista) [inline core/web/evm_forwarders_controller.go:96]: I don't know much about the forwarder, but if we didn't support multiple chains before... how did it work, you could only use forwarders on the default chain or something?

Comment (krehermann) [reply core/web/evm_forwarders_controller.go:96]: no, we were leaking the DB id and expecting user to use that. that DB id was unique b/c all the forwarders were in one table

File: core/store/migrate/plugins/relayer/evm/testutils/db.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		panic(fmt.Sprintf("failed to exec SQL for the initial state of the EVM database: %v", err))
+	}
+
+}
+
+//go:embed evm_initial_state.sql
+var evmInitialState string
+
+// NewEVMDB creates a new EVM database with the given schema and chainId
+func NewDB(t testing.TB, cfg evm.Cfg) *sqlx.DB {
```
Comment (reductionista) [inline core/store/migrate/plugins/relayer/evm/testutils/db.go:64]: Name of function in comment should match name of function

File: core/services/relay/evm/evm.go [hunk 5]
```diff
@@ -128,12 +149,27 @@ func NewRelayer(lggr logger.Logger, chain legacyevm.Chain, opts RelayerOpts) (*R
...
+	// run the migrations for the relayer
+	db, err := sqlx.Open(string(dialects.Postgres), opts.DBURL.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to open db: %w", err)
+	}
+
+	err = evmdb.Migrate(context.Background(), db, evmdb.Cfg{
+		Schema:  "evm_" + chain.ID().String(),
+		ChainID: ubig.New(chain.ID()),
+	})
```
Comment (reductionista) [inline core/services/relay/evm/evm.go:161]: Have we thought through what happens if the node gets shut down while the relay is migrating? Is completing the migration what we want, or would it be better to abort?

File: core/store/migrate/plugins/relayer/evm/0002_initial.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+		return fmt.Errorf("failed to resolve down sql: %w", err)
+	}
+	downFunc := func(ctx context.Context, tx *sql.Tx) error {
+		_, err := tx.ExecContext(ctx, downSQL.String())
+		return err
+	}
+	goose.AddMigrationContext(upFunc, downFunc)
+	return nil
+}
+*/
```
Comment (reductionista) [inline core/store/migrate/plugins/relayer/evm/0002_initial.go]: TODO: remove if unused

File: core/store/migrate/plugins/relayer/evm/resolver.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	return nil
+}
+
+func generateMigrations(fsys fs.FS, rootDir string, tmpDir string, val Cfg) ([]string, error) {
+	err := os.MkdirAll(tmpDir, os.ModePerm)
+	if err != nil {
+		return nil, err
+	}
+	b := make([]byte, 1024*1024)
```
Comment (reductionista) [inline core/store/migrate/plugins/relayer/evm/resolver.go:61]: I'm not sure how to do this cleanly off the top of my head, but I assume once we get past the initial draft stage, we'd want to make this able to handle arbitrarily large migration files. If not, we should at least make it a constant or config option so we can increase it later if needed.

File: core/store/migrate/plugins/relayer/evm/initUp.tmpl.sql [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+	receipt jsonb NOT NULL,
+	created_at timestamptz NOT NULL,
+	CONSTRAINT chk_hash_length CHECK (((octet_length(tx_hash) = 32) AND (octet_length(block_hash) = 32))),
+	CONSTRAINT eth_receipts_pkey PRIMARY KEY (id),
+	CONSTRAINT eth_receipts_tx_hash_fkey FOREIGN KEY (tx_hash) REFERENCES {{ .Schema }}.tx_attempts(hash) ON DELETE CASCADE
+);
+CREATE INDEX idx_eth_receipts_block_number ON {{ .Schema }}.receipts USING btree (block_number);
+CREATE INDEX idx_eth_receipts_created_at ON {{ .Schema }}.receipts USING brin (created_at);
+CREATE UNIQUE INDEX idx_eth_receipts_unique ON {{ .Schema }}.receipts USING btree (tx_hash, block_hash);
+*/
```
Comment (reductionista) [inline core/store/migrate/plugins/relayer/evm/initUp.tmpl.sql:295]: Is this uncommented code all WIP?

File: core/store/migrate/plugins/relayer/evm/migrate.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+	goose.SetTableName(fmt.Sprintf("goose_migration_relayer_%s_%s", cfg.Schema, cfg.ChainID.String()))
+	err := Register0002(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to register migration 0002: %w", err)
+	}
+	return nil
+}
+*/
+// Migrate migrates a subsystem of the chainlink database.
+// It generates migrations based on the template for the subsystem and applies them to the database.
```
Comment (reductionista) [inline core/store/migrate/plugins/relayer/evm/migrate.go:89]: TODO: update comment to be evm specific (I assume "chainlink database" will no longer be a meaningful concept, right?)

File: core/store/migrate/plugins/relayer/evm/migrate.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+		err = os.MkdirAll(d, os.ModePerm)
+		if err != nil {
+			return fmt.Errorf("failed to create directory %s: %w", d, err)
+		}
+		migrations, err := generateMigrations(embeddedTmplFS, MigrationRootDir, d, cfg)
+		if err != nil {
+			return fmt.Errorf("failed to generate migrations for opt %v: %w", cfg, err)
+		}
+		fmt.Printf("Generated migrations: %v\n", migrations)
+	*/
```
Comment (reductionista) [inline core/store/migrate/plugins/relayer/evm/migrate.go]: Getting a little confused about all of the commented code.  I see `setupPluginMigraitons is commented out, as well as `Register0002`, then this portion of `Migrate` which calls `generateMigrations` is commented, even though `generateMigrations` is uncommented? Which of these will we end up using, or are they all still WIP?

File: core/services/chainlink/relayer_factory.go [hunk 11]
```diff
@@ -72,6 +74,7 @@ func (r *RelayerFactory) NewEVM(ctx context.Context, config EVMFactoryConfig) (m
 
 		relayerOpts := evmrelay.RelayerOpts{
 			DS:                   ccOpts.DS,
+			DBURL:                &dburl,
```
Comment (reductionista) [inline core/services/chainlink/relayer_factory.go:77]: Cool, so... we tell all relays to use the databse url in the global toml config (`config.AppConfig.Database().URL()`) by default, but allow it to be overridden on a per-relay basis by reading the `CL_DATABASE` environment variable that's set locally on each running relay? I think that makes sense

**Comment by reductionista** [comment]
I don't see any big problems with this approach, seems like some good progress in the right direction. Left some minor comments and questions, but most of them may not be too relevant at this stage as it's such an early draft--I assume more of the details will be filled in as it progresses.

**Pull request closed without merging**